### PR TITLE
Tao 2002 auto termination after pause

### PR DESCRIPTION
**Pull request by Alroniks**
https://oat-sa.atlassian.net/browse/TAO-2002

Related to https://github.com/oat-sa/extension-tao-act/pull/186

### CODE REVIEW

**Comment by Alroniks** [comment]
@krampstudio Please, take a look also.

File: model/implementation/DeliveryExecutionStateService.php [hunk 1]
```diff
@@ -88,6 +92,32 @@ public function waitExecution(DeliveryExecution $deliveryExecution)
...
+                if ($wasPausedAt && $this->hasOption('termination_delay_after_pause')) {
+                    $delay = $this->getOption('termination_delay_after_pause');
+                    if ($wasPausedAt->add(new DateInterval($delay)) < (new DateTime())) {
+                        $this->terminateExecution($deliveryExecution, null, ['test' => TestSessionProcessor::TEST_CODE_INCOMPLETE]);
+
+                        return false;
+                    }
+                }
+            }
+
```
Comment (hutnikau) [inline model/implementation/DeliveryExecutionStateService.php]: Could you move all this stuff from `waitExecution()` function to separate protected method (something like `isExpired()`)?

**Comment by krampstudio** [comment]
Could you just explain me your approach. The goal is terminate executions paused from some amout of time, so what does trigger the given event ?

File: model/implementation/DeliveryExecutionStateService.php [hunk 2]
```diff
@@ -20,11 +20,15 @@
 
 namespace oat\taoProctoring\model\implementation;
 
+use DateInterval;
+use DateTime;
+use DateTimeImmutable;
 use oat\oatbox\service\ConfigurableService;
-use oat\taoQtiTest\models\TestSessionMetaData;
+use oat\taoAct\model\testSession\TestSessionProcessor;
```
Comment (boajer) [inline model/implementation/DeliveryExecutionStateService.php]: Is it necessary to introduce new dependency from act extension?

Comment (krampstudio) [inline model/implementation/DeliveryExecutionStateService.php]: Good catch!

Comment (hutnikau) [inline model/implementation/DeliveryExecutionStateService.php]: My opinion is that it is not necessary to add new dependency and use `TestSessionProcessor` here. You should not pass exit codes to the `DeliveryExecutionTerminated` event. You can implement new custom event (something like `DeliveryExecutionExpired`) and add listener of this event to the `TestSessionProcessor`. The reason of creation `TestSessionProcessor` was to let him decide for themselves which exit code should be saved, and encapsulate all work with exit codes in it.

**Comment by Alroniks** [comment]
Closed per favor https://github.com/oat-sa/extension-tao-proctoring/pull/108

**Pull request closed without merging**